Task 1.2 (a) Move Development Code related to D2

xandie985 opened the pull request:
### **D2 Layer Implementation for Time Series Dataset**  

#### **Key Changes**  
1. **Architecture**  
   - [x] Split into modules:  
     - `time_series_d2.py`: Model-ready processing (`TSDataModule`)  

2. **D2 Layer** (`TSDataModule`)  
   - [ ] PyTorch Lightning integration  
   - [ ] Sliding window generation with NaN-aware indexing  
   - [ ] Flexible train/val/test splits  

3. **Code Quality**  
   - [ ] Docstrings, type hints, error handling, logging  

4. **Testing**  
   - [x] Unit tests for D2 layers

agobbifbk commented:
A couple of comments:
- move the D1/D2 layer into folder (we can have multiple d1 and d2) and give them proper names

fkiraly commented:
reminder: do not forget to merge `main` into this and resolve conflicts

xandie985 commented:
Hi @fkiraly, this PR is ready for review. Can you provide your valuable feedback on this?

agobbifbk commented:
One first comment: 
IMHO the structure should be:
├data_structure
├──d1_layers
├────utils.py
├────MultiSourceCSV.py
├────otherSources.py 
├──d1_layers
├────utils.py
├────encoderdecoder.py
├────otherd2layer.py

agobbifbk commented:
Second comment:
this should me moved into the test part for cleaner code
https://github.com/DSIP-FBK/DSIPTS_PTF/blob/b7213be79608d79b3849490355940c05e73a1663/dsipts/data_structure/time_series_d1.py#L51

agobbifbk commented:
Other comment:
here https://github.com/DSIP-FBK/DSIPTS_PTF/blob/b7213be79608d79b3849490355940c05e73a1663/dsipts/data_structure/time_series_d1.py#L693 we need to tell to the user that we are inferring the frequency from the data (logging? print?)

agobbifbk commented:
I would like to keep also the temporal categorical variables relate to time as in the actual implementation :
```python
    def enrich(self,dataset,columns):
        if columns =='hour':
            dataset[columns] = dataset.time.dt.hour
        elif columns=='dow':
            dataset[columns] = dataset.time.dt.weekday
        elif columns=='month':
            dataset[columns] = dataset.time.dt.month
        elif columns=='minute':
            dataset[columns] = dataset.time.dt.minute
```
some comments:
- the user should be able to define `enrich_cat ` as a list of the  strings you see above
- it should be done only if the temporal columns is a timestamp
- the `enrich_cat ` should be added to the `cat_cols`  list
- the `enrich_cat ` should be added to the `known_cols`  list

implementation speaking the easiest solution can be to add the `enrich` after every `load.csv` instance, maybe there is a smarter way to do it but this approach will not affect the rest of the code.

agobbifbk commented:
Related to D2:
- TSDataModule should be renamed something like EncoderDecoder
- in the __init__ method: 
```python
d1_dataset: MultiSourceTSDataSet, 
```
this should be something related to ANY d1 layer --> maybe is better to create a general class d1 and the current `MultiSourceTSDataSet`  is an extension. We can use the base class to show to the user which methods are mandatory (basically  `__getitem__` and  `__len__` and other fields)

-  again in the init:
```python
        known_cols: Optional[List[str]] = None,
        unknown_cols: Optional[List[str]] = None,
```
 these can be extracted from the d1 layer if I'm not wrong.
 -  TSDataModule should not have `__getitem__` and  `__len__`, these are only for the Dataset. I suggest to extend the pytorch  `Dataset` class with all the logic in the current `__getitem__` and  `__len__` you wrote for TSDataModule

agobbifbk commented:
Ok I think we are converging, great! A couple of questions:
1- do you think we can move this block to the base d1 class:
```python
    @property
    def group_cols(self) -> List[str]:
        """Get the group columns."""
        return self._group_cols

    @property
    def target_cols(self) -> List[str]:
        """Get the target columns."""
        return self._target_cols

    @property
    def feature_cols(self) -> List[str]:
        """Get the feature columns."""
        return self._feature_cols

    @property
    def cat_cols(self) -> Optional[List[str]]:
        """Get the categorical columns."""
        return self._cat_cols

    @property
    def known_cols(self) -> Optional[List[str]]:
        """Get the known future columns."""
        return self._known_cols

    @property
    def unknown_cols(self) -> Optional[List[str]]:
        """Get the unknown future columns."""
        return self._unknown_cols
```

Is this used in the d2 layer? 
```
        categorical_encoders: Optional[Dict] = None,

```
I see that `__getitem__` and `__len__` are still in the D2 layer this should be moved into TimeSeriesSubset (and rename it as EncoderDeconderDataset or something like that). All the logic about windowing should be inside the dataset while the D2 manage the creation of the indexes, the train/validation/test split procedure

agobbifbk commented:
I see another minor point here
```python
 known_cols: Optional[List[str]] = None,
 unknown_cols: Optional[List[str]] = None,
``` 
are still in the init of the d2 (should be inherited from d1)

agobbifbk commented:
I see another strange thing:
```python
    # Check if this is a test case and handle it separately
    test_result = (df)
```
 Why we call `extend_time_df_test_case`  also in the case we want to use the method? I suggest to write your csv test in the test folder and load it when you are doing tests, this seems more reasonable IMHO

agobbifbk commented:
I'm actually using d1/d2 layer:
1- I got this message:
2025-07-25 08:39:01,676 - __main__ - WARNING - Time column 'date' is not datetime type. Temporal enrichment skipped.
usually when you load a csv the temporal column is encoded as string, try something like this
```python
        df = pd.read_csv(file_path)
        try:
            df[self.time_col] = pd.to_datetime(df[self.time_col])
        except:
            logging.info(f'Can not parse {self.time_col} to datetime')
```
2- this part fails if groups is None or empty list:
```python
for group_key, group_data in chunk.groupby(self.group_cols):
        # Convert single values to tuples for consistency
        if not isinstance(group_key, tuple):
            group_key = (group_key,)
```

agobbifbk commented:
Maybe something like:
```python
        self.use_groups = True
        if group_cols is None:
            group_cols = '_GROUP_'
            self.use_groups = False
        
        if isinstance(group_cols, str):
            self._group_cols = [group_cols]
        else:
            self._group_cols = group_cols
``` in the init and then:
```python
  for chunk in pd.read_csv(file_path, chunksize=self.chunk_size):
      if not self.use_groups:
         chunk['_GROUP_'] = 0 
``` after each reading  phase should fix it

I would like then to have this information in the metadata since the groups are used as categorical variable in case of non global forercaster

agobbifbk commented:
Another comment, it seems that the metadata from d1 layer are missing in the current version or at least a function is defined:
```python
    def _prepare_metadata(self):
        """
        Prepare metadata for efficient data access.
        """
        # Create a cumulative index mapping for efficient lookup
        self.cumulative_lengths = [0]
        for file_group_key in self._group_ids:
            group_length = self.group_info[file_group_key]["length"]
            self.cumulative_lengths.append(self.cumulative_lengths[-1] + group_length)

        # Store total dataset length
        self.dataset_length = self.cumulative_lengths[-1]

        logger.info(f"Dataset prepared with {self.dataset_length} total samples")
``` but it does not create any metadata dict, can you check it? 
we already discussed about this it should contain information such as:
- the dimension of the target variables
- the dimension of past numerical and categorical variables
- the dimension of the future numerical and categorical variables

Can you check please what we are doing in PTF? 
THX

agobbifbk commented:
Hello thee, I see a lot of useless duplication:
```python
        x = {
            # Encoder data
            "encoder_cont": encoder_cont,
            "encoder_cat": encoder_cat,#PUT IT ONLY IF PRESENT
            "encoder_lengths": encoder_lengths, #USELESS
            "encoder_time_idx": encoder_time_idx,#USELESS
            "encoder_mask": encoder_mask,#USELESS
            # Decoder data
            "decoder_cont": decoder_cont,#PUT IT ONLY IF PRESENT
            "decoder_cat": decoder_cat,#PUT IT ONLY IF PRESENT
            "decoder_lengths": decoder_lengths,#USELESS
            "decoder_time_idx": decoder_time_idx,#USELESS
            "decoder_mask": decoder_mask,#USELESS
            # Static features
            "static_categorical_features": static_categorical_features,
            "static_continuous_features": static_continuous_features,
            # Target information
            "idx_target": idx_target,
            "target_scale": target_scale,#USELESS
            # Model-compatible keys
            "x_num_past": encoder_cont,
            "x_cat_past": encoder_cat,
            "x_num_future": decoder_cont,
            "x_cat_future": decoder_cat,
            "y": future_targets,
            # Backward compatibility keys
            "past_features": encoder_cont,
            "future_targets": future_targets,
            # Metadata
            "group_id": window["group_id"],
            "past_time": [past_samples[i]["past_time"] for i in range(len(past_samples))],#USELESS
            "future_time": [future_samples[i]["future_time"] for i in range(len(future_samples))],#USELESS
        }
``` 
This structure is not what we agreed at the beginning (please see the comments and check duplicates)


moreover these lines can confuse the user:
```python
encoder_cat = torch.zeros(self.past_len, len(self.cat_feature_cols))  # Placeholder

# Future features (known at prediction time)
decoder_cont = torch.zeros(self.future_len, len(self.cont_feature_cols))  # Placeholder
decoder_cat = torch.zeros(self.future_len, len(self.cat_feature_cols))  # Placeholder
```
if these variable are not present in the dataset they should not be present in the batch structure (aka keys of the dictionary)

xandie985 commented on dsipts/data_structure/time_series_d2.py:
```diff
@@ -470,45 +535,134 @@ def __getitem__(self, idx):
+        # Create decoder features (future) - only known features would be available
+        # For now, assume no future features are known
+        decoder_cat = torch.zeros((self.future_len, 0))
+        decoder_cont = torch.zeros((self.future_len, 0))
```
TODO: Remove the categorical keys when there are no `cat` keys present (the dataset doesn't have categorical values), instead of having null/empty column, we DO NOT want to store the cat key at all.
Also, we don't initialize the `decoder_cat  `and `decoder_cont`.

xandie985 commented:
feat: robust time handling + D2 index maps + optional target in decoder

- **auto-create `time` col if missing**; coerce existing cols to datetime/numeric; we always sort by time
- **d2 now exposes column indices:** `idx_known_num`, `idx_unknown_num`, `idx_known_cat`, `idx_unknown_cat`, `idx_target`
- **add `include_target_in_decoder` flag**; when True supplies `decoder_target`

xandie985 commented:
Key changes in D2after D1 Improvments. 

**Issues & Improvements**  

1. **using D1 metadata as single source of truth**  
   - **Issue**: recomputing of indices in `getitem` is brittle.  
   - **Fix**: Directly use D1 metadata indices (`feature_cols`, `idx_categorical`, `idx_known_future`, `idx_targets`).  
   - **Benefit**: Ensures consistency with D1 ordering

2. **Properly split numeric/categorical tensors**  
   - **Issue**: `x_num_past` incorrectly includes categoricals; placeholders were being used
   - **Fix**: Slice `X` using `idx_num` (non-categorical indices) and `idx_categorical`.  
   - **Benefit**: Real data flows through; correct types (`float32` for numeric, `int64` for categorical).  

3. **Remap `idx_target` relative to `x_num_past`**  
   - **Issue**: `idx_target` points to full `X` indices, not numeric-only.  
   - **Fix**: Map `idx_target` to numeric-only spacee.  

4. **correct categorical cardinality**  
   - **Issue**: it uses `.categories_[0]` instead of `.classes_`
   - **Fix**: `cardinality = len(encoder.classes_)`
   - **Ensure**: Cardinalities match `x_cat_past` column order (use `categorical_mappings`)

---
TODOs:
- Update the test cases to test the implementation 
- Ensure the test cases pass and increase coverage after discussion with @agobbifbk

agobbifbk commented:
Some comments:
1 -   'idx_known_num', 'idx_unknown_num', 'idx_known_cat', 'idx_unknown_cat', 'categorical_cardinality_past' , 'past_features' , 'time_idx' are not needed in the batch
2 - time_idx can be used for debugging: for example use `x["time_idx"] = start_idx` so we can check with the original csv 
3- in the batch `group_id` seems a string not the id of the group and it is not needed in the batch (BUT KEEP IT FOR DEBUGGING)
4- the logic of groups is missing here. The D1 layer should accept a boolean parameter called `global_forecasting` : if true nothing need to be changed, if it is `False` AND there are more than 1 group, we need to add the group to the *categorical known* variables and return them also in the batches (but if you this correctly in the D1, the D2 should provide the correct data
5- the D1 layer should apply the label encoder also to the groups, especially if global encoder is `False`

xandie985 commented:
Updates:

**1. Streamlined Batch Keys**  
- Removed non-essential metadata/debug keys (`idx_known_num`, `idx_unknown_num`, etc.)  
- Retained only training-critical keys → cleaner batches, lower memory  
- Verified via unit tests  

**2. Enhanced Debugging with `time_idx`**  
- Replaced counter with actual data start index → better traceability to source CSV  
- Implementation: `x["time_idx"] = start_idx`   (foe debugging)

**3. Consistent `group_id` Handling**  
- enforced integer type (converted strings via metadata mapping)  
- Added validation tests  

**4. Flexible Forecasting Modes**  
- Default: `global_forecasting=False` (D1 layer)  
- Auto-converts groups → categorical vars for local forecasting  
- ttested both global/local scenarios  (improvements in testing needed here)

**5. Robust Categorical encoding**  
- Groups encoded via label encoder when `global_forecasting=False`  
- Safety checks for temporal features (eg. "hour", "minute") → enforced categorical treatment

agobbifbk commented:
1- if global forecaster is False (default) the groups should appear in the categorical variables (D1) and the number of classes is missing too
2- there is still an issue: with minutes as categorical I see in the metadata that n_categorical is false,
3 -future_targets in D2 is the same as y
4- but in the D2 when there are not groups

xandie985 commented:
**The previous commit contains the following enhancements:**

✅ Added StandardScaler functionality to EncoderDecoder class
✅ Implemented support for custom scalers
✅ Added option to scale target variables
✅ Ensured scalers are fitted only on training data
✅ Updated split method to apply scalers after splitting
✅ Created comprehensive tests for the scaling functionality

xandie985 commented:
- `future_targets `and `y` denote the same stuff; `future_targets`  they will be removed in the next push. 
- `minutes `will be transferred into the `x_cat_future` now (all the temporal enhancement columns by default) + known categorical column values as well

- `x_cat_future `will have the following logic now:
If only temporal enhancement (TE): only those values will be populated in the `x_cat_future`
If only known columns: it's checked if it's categorical, then added to `x_cat_future`
If no TE and no known columns given: empty `x_cat_future`
If both TE and known columns: add both in `x_cat_future`

- `group_id` will be removed in the next push.

xandie985 commented:
**Batch Processing Updates (D2 Layer Output)**
---
### Key Improvements:
- **Unified Batch Structure**: Standardized D2 layer output to always return a tuple `(x, y)` where:
  - `x` contains all input features and metadata
  - `y` contains target values
- **Model-Compatible Keys**: Added model-compatible keys in batch output:
  ```python
  {
      'x_num_past': Past numeric features,
      'x_cat_past': Past categorical features,
      'x_cat_future': Future known categorical features,
      'y': Target values,
      'idx_target': Target indices,
      'group_id': Group identifier (for global_forecasting=True),
      'time_idx': Time index for debugging
  }
  ```
- **Temporal Feature Handling**: Temporal features (hour, dow, month) are now properly included in categorical tensors
- **Global vs Local Forecasting**:
  - `global_forecasting=True`: Includes `group_id` in batch
  - `global_forecasting=False`: Group information embedded in `x_cat_past`/`x_cat_future`

### **Min-Max Scaling Enhancements**

Key Features:
- **Robust Implementation**:
  - Handles edge cases (constant features, NaN values)
  - Supports both single and multi-feature scaling
  - Preserves data distribution while normalizing to [0,1] range

Usage:
```python
from dsipts.data_structure.scalers import MinMaxScaler

# Initialize and fit
scaler = MinMaxScaler()
scaler.fit(data)

# Transform data
scaled_data = scaler.transform(data)

# Inverse transform
original_data = scaler.inverse_transform(scaled_data)
```

### **Standard Scaling Implementation**

New Features:
- **Z-Score Normalization**: Implements standard scaling (mean=0, std=1)
- **Efficient Computation**:
  - Handles sparse matrices efficiently
  - Supports partial fitting for large datasets
  - Maintains numerical stability with epsilon term

Usage:
```python
from dsipts.data_structure.scalers import StandardScaler

# Initialize and fit
scaler = StandardScaler()
scaler.fit(data)

# Transform data
scaled_data = scaler.transform(data)

# Inverse transform
original_data = scaler.inverse_transform(scaled_data)
```

### **Integration Between Components**

- **Seamless D1→D2 Integration**: Both scaling methods work transparently between D1 and D2 layers
- **Metadata Preservation**: Scaling parameters are preserved and transferred between layers
- **Batch Processing**: Both scaling methods work efficiently in batch processing pipelines

### **Testing & Validation**

- Comprehensive test coverage for all scaling operations
- Validation of inverse transform accuracy
- Performance testing for large-scale datasets
- Integration tests for D1→D2 workflow with both scaling methods

xandie985 commented:
Last commit:

### refactor: Isolate scaling logic into dedicated `scalers.py` module

This commit completes the refactoring of all data scaling functionality. The core logic has been moved from the D2 layer (`encoder_decoder.py`) into a new, self-contained `scalers.py` module. This creates a cleaner, more modular, and more maintainable architecture.

The `EncoderDecoder` class now delegates all scaling operations to the `ManualScaler` and no longer contains any scaling implementation details. The public API remains unchanged, ensuring no breaking changes for downstream users.

### Key Changes

* **New `scalers.py` Module ✨**
    * Introduced `ManualScaler`, a new class that encapsulates all scaling logic (fitting, transforming, and inverse transforming).
    * It internally uses `StatsRecorder` and `MinMaxRecorder` to incrementally compute the required metadata (mean, std, min, max) when operating in memory-efficient mode.
    * This implementation is dependency-free and does not require `sklearn`.

* **`encoder_decoder.py` Updates 🔄**
    * Removed ~200 lines of scaling code, significantly simplifying the D2 layer.
    * The `EncoderDecoder` now initializes and uses `ManualScaler`.
    * Delegate methods (`fit_scaler`, `transform_with_scaler`, `apply_inverse_scaling`) are preserved to maintain the existing API.

* **Memory Efficiency Logic**
    * The module respects the `memory_efficient` flag from the D1 layer.
    * If `False`, metadata is calculated by loading the training data into memory at once.
    * If `True`, metadata is calculated incrementally by iterating through the training data in chunks, avoiding high memory usage.

### Status

This is a checkpoint commit for the refactored implementation. Initial integration tests for standard and min-max scaling are successful. 
- [x] Thorough testing across all scenarios for scalers.
- [x] Reduce the D1 length, make it more modular. It's too big now. 
- [x] Category cols and their cardinalities should be RELATIVE and in separate LIST (remove the current dictionary data structure)
- [x] Switch from known_cols, unknown cols to past_cols and future_cols. 
  - Current implementation assumes known value to be placed in both future and past, and unknown only in past. But if want to keep a known value only in future but not in past?
  - For this scenario there is need for future and past cols that handles such situation with more clarity.

xandie985 commented:
### D1 Layer

* **Standardized Feature Columns:** weve shifted the data model from `known`/`unknown` columns to the more explicit **`past`/`future` paradigm**.
* **Flexible Data Loading:** Implemented `read_options` to allow users to pass custom arguments (e.g., `sep`, `na_values`, `usecols`) to the underlying [pd reader functions](https://pandas.pydata.org/docs/reference/api/pandas.read_csv.html)
* **Unified Data Ingestion:** The `_data_chunk_generator` and `_process_data` functions are unified to handle multiple file formats, including **CSV, Parquet, JSON, and Pickle**, from a single entry point.
* **Consistent Feature Enrichment:** temporal enrichment columns (`enrich_cat`) are now consistently added to both **`past_cols`** and **`future_cols`**.
* **Centralized Metadata:** calculation for categorical cardinality has been updated and is now part of the **`D1.metadata`** object.

***
### D2 Layer (mojor rework on scaler here)

* **Unified On-the-Fly Scaling:** scaling transformation logic is now centralized in **`EncoderDecoderDataset.__getitem__`**. This single implementation applies scaling on-the-fly for **both `memory_efficient=True` and `False` modes**, eliminating duplicate code.
* **Efficient Batched Fitting:** The **`fit_scaler`** method now uses a **`torch.DataLoader`** (with a batch size of 64) to process the training set in efficient chunks. This replaces the previous sample-by-sample `for` loop and is used for both `partial_fit` and `fit` operations.
* **Correct Invocation:** The scaler fitting process is correctly invoked via the **`_apply_scaling_to_splits`** method, which ensures the scaler is fitted *only* on the training data to prevent data leakage.

***
### Refactoring & Code Hygiene

* **Decoupled Architecture:** The `d2_layer` cross-reference was removed, **decoupling** the `EncoderDecoderDataset` from its parent `DataModule`. The dataset now receives the fitted scalers directly.
* **Simplified Workflow:** The redundant methods and the wrapper classes were completely removed for the unified on-the-fly approach.
* **Removed Obsolete Code:** The legacy **`dsipts/data_structure/scalers.py`** file was deleted, removing an entire module that is no longer needed.

***
### Test Cases

* **Expanded Test Suite:** The test suite was successfully updated and passed, with a total of **39 passing tests**.
* **Comprehensive Logging:** A new test, **`test_d1_d2_analysis`**, was added to generate detailed logs for manual verification of the pipeline's behavior across all major configurations.

xandie985 commented:
**D2 auto scaling, opt-in decoder targets & leaner D1 samples**  

**Summary of changse:**  
- D2 now fits scalers **only on train split** automatically when you pass `split_config`.  
- Two new flags give you full control over where the target column shows up:  
  – `include_target_in_decoder` → feeds it to the decoder input (teacher-forcing style).  
  – `add_target_to_past` → adds it to the past features for autoregressive models.  
- `D1.__getitem__` no longer spits out `cat_cols` / `cat_cardinalities` in every sample; grab them once from `d1.metadata`.  
- Tests & helpers updated to match the new flow; no breaking changes.

**One-liner for scaler implementation**  
```python
d2 = EncoderDecoder(d1, past=24, future=12, scaling="standard",
                    split_config=(0.7,0.15,0.15))   # scaler is fitted
```

xandie985 commented:
### 🔧 Refactor: EncoderDecoder DataModule Overhaul

**Summary:** Major refactor aligning with Lightning lifecycle patterns, fixing data leaks, and optimizing performance.

**Key Changes:**

- **Lifecycle Alignment**: Replaced custom `setup()` with standard `setup(stage)` + stateful flags for idempotent operations
- **Data Leak Fix**: Added validation to prevent group overlaps between train/val/test splits
- **Pure Scaling**: Scalers now fit only on training window timesteps, preventing future data leakage
- **Performance**: Added LRU caching for on-the-fly transforms + fixed pre-computation bugs
- **Robustness**: Unified splitting logic with clear global/local forecasting behavior

xandie985 commented:
### 🎯 Categorical Feature Optimization + Tensor Format Consistency across all features (BLF)

### Key Improvements maje:
- **5-10x performance boost**: Pre-calculate indices once instead of recalculating per-sample
- **~30-40% memory savings**: Consolidated redundant arrays into single tensors with efficient slicing
- **Cleaner code**: `_get_window_no_cache` reduced from 150→50 lines
- **Single source of truth**: D1 metadata drives all index operations

### Fixed these:
✅ **BLF Format Consistency** — All tensors follow (Batch, Length, Features) or (Length, Features)  
✅ **Cardinality Mapping** — Correct D1→D2 index mapping via column name lookup  
✅ **Categorical Order** — Validated across indices, names, and cardinalities  
✅ **Index Pre-Calculation** — `idx_cat_past`, `idx_cat_future`, `idx_num`, `idx_num_future` computed once

### Data Flow 
1. **D1 Layer**: Label-encodes categorical columns, stores indices & cardinalities in metadata
2. **D2 `__init__`**: Pre-calculates index lists from D1 metadata (*once*)
3. **D2 `__getitem__`**: Uses pre-calculated indices for efficient tensor slicing
4. **Collation**: DataLoader batches individual samples
5. **Model Input**: Final batch with `x_cat_past` (Batch, Length, Num_Cat) and `x_cat_future` ready for embeddings

Current format of the data:
```python
for key in batch1.keys():
    print(key, '\t', batch1[key].shape)
```
```bash
x_num_past 	          torch.Size([32, 96, 21])
x_cat_past 	          torch.Size([32, 96, 3])
x_cat_future 	          torch.Size([32, 96, 2])
y 	  	  	          torch.Size([32, 96, 1])
idx_target 	          torch.Size([32, 1])
time_idx 	  	          torch.Size([32])
```

xandie985 commented:
The scaling now happens two-way:
1. if there are gaps in the windows, we use the masks. 
2. other wise we treat the data group-wise. 

chunk_size has more functionality now. 

Performed some basic cleanups + test case updates
